reactor-core/src/main/java/reactor/core/publisher/SinksSubscribers.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2011-Present VMware Inc. or its affiliates, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.core.publisher;
+
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import org.reactivestreams.Subscription;
+
+import reactor.core.CoreSubscriber;
+import reactor.core.Disposable;
+import reactor.util.context.Context;
+
+/**
+ * @author Simon Baslé
+ */
+final class SinksSubscribers {
+
+	private SinksSubscribers() {}
+
+	static abstract class AbstractSubscriberAdapter<T, SINK> implements CoreSubscriber<T>, Disposable {
+
+		final SINK sink;
+		final Context context;
+		final Sinks.EmitFailureHandler emitFailureHandler;
+
+		volatile Subscription                                                     s;
+		@SuppressWarnings("rawtypes")
+		static   AtomicReferenceFieldUpdater<AbstractSubscriberAdapter, Subscription> S
+				= AtomicReferenceFieldUpdater.newUpdater(AbstractSubscriberAdapter.class, Subscription.class, "s");
+
+		AbstractSubscriberAdapter(SINK sink) {
+			this.sink = sink;
+			this.context = sink instanceof ContextHolder ? ((ContextHolder) sink).currentContext() : Context.empty();
+			this.emitFailureHandler = Sinks.EmitFailureHandler.FAIL_FAST;
+		}
+
+		@Override
+		public Context currentContext() {
+			return context;
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (Operators.setOnce(S, this, s)) {
+				s.request(Long.MAX_VALUE);
+			}
+		}
+
+		@Override
+		public boolean isDisposed() {
+			return s == Operators.cancelledSubscription();
+		}
+
+		@Override
+		public void dispose() {
+			Operators.terminate(S, this);
+		}
+	}
+
+	static final class ManySubscriberAdapter<T> extends AbstractSubscriberAdapter<T, Sinks.Many<T>> {
+
+		ManySubscriberAdapter(Sinks.Many<T> many) {
+			super(many);
+		}
+
+		@Override
+		public void onNext(T t) {
+			sink.emitNext(t, emitFailureHandler);
```

Comment:
since some `EmitFailureHandler`s can be blocking (we do allow that), using `emitNext` here is dangerous and error-prone

Reply:
although if we're talking about a `Subscriber` usage, most failure codes should be intrinsically avoided by the mere fact that we're within Reactive Streams guarantees. that said, it shouldn't be too hard to switch to `tryEmit` API.